libs/parallelism/executors/include/hpx/executors/thread_pool_scheduler_bulk.hpp

```diff
@@ -0,0 +1,301 @@
+                std::atomic<decltype(hpx::util::size(shape))> tasks_remaining{
+                    hpx::util::size(shape)};
+                hpx::util::detail::prepend_t<
+                    value_types<hpx::tuple, hpx::variant>, hpx::monostate>
+                    ts;
+                std::atomic<bool> exception_thrown{false};
+                std::optional<std::exception_ptr> exception;
+
+                template <typename Sender_, typename Shape_, typename F_,
+                    typename Receiver_>
+                operation_state(thread_pool_scheduler&& scheduler,
+                    Sender_&& sender, Shape_&& shape, F_&& f,
+                    Receiver_&& receiver)
+                  : scheduler(std::move(scheduler))
+                  , op_state(hpx::execution::experimental::connect(
+                        std::forward<Sender_>(sender), bulk_receiver{this}))
+                  , shape(std::forward<Shape_>(shape))
+                  , f(std::forward<F_>(f))
+                  , receiver(std::forward<Receiver_>(receiver))
+                {
+                }
+
+                void start() noexcept
+                {
+                    op_state.start();
+                }
+            };
+
+            template <typename Receiver>
+            auto connect(Receiver&& receiver) &&
+            {
+                return operation_state<std::decay_t<Receiver>>{
+                    std::move(scheduler), std::move(sender), std::move(shape),
+                    std::move(f), std::forward<Receiver>(receiver)};
+            }
+
+            template <typename Receiver>
+            auto connect(Receiver&& receiver) &
+            {
+                return operation_state<std::decay_t<Receiver>>{scheduler,
+                    sender, shape, f, std::forward<Receiver>(receiver)};
+            }
+        };
+    }    // namespace detail
+
+    template <typename Sender, typename Shape, typename F,
+        HPX_CONCEPT_REQUIRES_(std::is_integral_v<std::decay_t<Shape>>)>
+    constexpr auto tag_dispatch(bulk_t, thread_pool_scheduler scheduler,
+        Sender&& sender, Shape&& shape, F&& f)
+    {
+        return detail::thread_pool_bulk_sender<std::decay_t<Sender>,
+            hpx::util::detail::counting_shape_type<std::decay_t<Shape>>,
+            std::decay_t<F>>{std::move(scheduler), std::forward<Sender>(sender),
+            hpx::util::detail::make_counting_shape(shape), std::forward<F>(f)};
+    }
+
+    template <typename Sender, typename Shape, typename F,
+        HPX_CONCEPT_REQUIRES_(!std::is_integral_v<std::decay_t<Shape>>)>
+    constexpr auto tag_dispatch(bulk_t, thread_pool_scheduler scheduler,
+        Sender&& sender, Shape&& shape, F&& f)
+    {
+        return detail::thread_pool_bulk_sender<std::decay_t<Sender>,
+            std::decay_t<Shape>, std::decay_t<F>>{std::move(scheduler),
+            std::forward<Sender>(sender), std::forward<Shape>(shape),
+            std::forward<F>(f)};
+    }
```

Comment:
Nice solution!

Reply:
Nah, I just liked the idea of creating a wrapper that implements the bulk part without replicating the rest.